Add fate-operator proposal. (#335)
kubeflow/community: proposals/fate-operator-proposal.md

-   [Background](#background)
-   [Motivation](#motivation)
-   [Goals](#goals)
-   [Non-Goals](#non-goals)
-   [Design](#design)
    -   [Container images and deploying FATE cluster on
        Kubernetes](#container-images-and-deploying-fate-cluster-on-kubernetes)
    -   [Custom Resource Definition](#custom-resource-definition)
        -   [Kubefate](#kubefate)
        -   [FATECluster](#fatecluster)
        -   [FateJob](#fatejob)
    -   [Controller design](#controller-design)

_Status_
* 2020-5-14 – Draft v1

## Background
Federated machine learning (FML) is a machine learning setting where many clients (e.g. mobile devices or organizations) collaboratively train a model under the coordination of a central server while keeping the training data decentralized. Only the encrypted mediate parameters are exchanged between clients with MPC or homomorphic encryption.
![Federated Machine Learning](diagrams/fate-operator-fl.png)

> FML has received significant interest recently, because of its effectiveness to solve data silos and data privacy preserving problems. Companies participated in federated machine learning include 4Paradigm, ANT Financial, Data Republic, Google, Huawei, Intel, JD.com, Microsoft, Nvidia, OpenMind, Pingan Technology, Sharemind, Tencent, VMware, Webank etc.

Reviewer: What do you mean by “participated” here?

Author: "Participant" in here means: join the research, and at least have own FML framework or wrapped existing frameworks as a solution. But not includes the companies just use FML or joined training. e.g. Googgle has Tensorflow Federated, OpenMide has PySyft, Tencent has solution based on FATE, Nvidia has Clara...

Reviewer: Given that there are so many frameworks there, would it possible to define some interface that can be generalized across federated learning frameworks?

Reviewer 2: Yeah, same idea here. Could we have something like fml-common?

Author: Indeed good idea. But FML is new, and there are many framework with different design purples. We would like to work with them and provide a common APIs. This is also why we hope to join Kubeflow community, that we can work together with other FML teams.

#### FATECluster
```
apiVersion: app.kubefate.net/v1beta1
kind: FateCluster
metadata:
  name: fatecluster-sample
  namespace: fate-9999
spec:
  version: v1.4.0
  partyId: "9999"
  proxyPort: "30009"
  partyList:
  - partyId: "10000"
    partyIp: "192.168.1.10"
    partyPort: "30010"
  egg:
replica: 1
  # Points to KubeFATE created by CRD “Kubefate”
  kubefate:
    name: kubefate-sample
    namespace:  kube-fate
```
The FateCluster defines a deployment of FATE on Kubernetes. 
* version defines the FATE version deployed in Kubernetes;
* partyId defines the FML party’s ID;
* proxyPort defines the exposed port for exchanging models and parameters between different parties in a FML training. It will be exposed as a node port;
* partyList defines the parties in a federation which take part in collaboratively learning;
* egg is the worker nodes of FATE.

#### FateJob
```
apiVersion: app.kubefate.net/v1beta1
kind: FateJob
metadata:
  name: fatejob-sample
  namespace: fate-9999
spec:
  # fateClusterRef points to the FATE cluster resource created by FateCluster
  fateClusterRef: fatecluster-sample

>   jobConf:

Reviewer: This looks very different from the existing spec of various operators in Kubeflow. Could you take a look at https://github.com/kubeflow/common to see if we can align the interface?

Author: We planned to reuse the common APIs, but a fundamental difference is hard to solve. To FATE, we need the FATE cluster is always deployed, and ready for training requests from both self party and collative parties, which means the training pod is not on demands.

          "guest": {
            "secure_add_example_0": {
              "seed": [
                123
              ]
            }
          },
          "host": {
            "secure_add_example_0": {
              "seed": [
                321
              ]
            }
          }
        },
        "algorithm_parameters": {
          "secure_add_example_0": {
            "partition": 10,
            "data_num": 1000
          }
        }
      }
```
FateJob defines the job sent to FATE cluster:
* fateClusterRef defines the cluster of FATE deployed on Kubernetes. Its value is resource name of FATE cluster created by CRD “FateCluster”;
* jobConf defines the details of a FML job. It includes:
   * pipeline: the workflow pipeline of FATE. In FATE, there are many prebuilt algorithm components (ref: https://github.com/FederatedAI/FATE/tree/master/federatedml and https://github.com/FederatedAI/FATE/tree/master/federatedrec), which can be used to train models. The pipeline defines how data are passed through and processed in the whole training flow;
   * moduleConf: the detail configuration of each algorithm component, e.g. the optimizers, the batch size etc.

> ### Controller design

Reviewer: Have you looked at https://github.com/kubeflow/common? Can we reuse the Kubeflow/common’s JobController interface ?

Reviewer 3: +1 on above comment

@terrytangyuan are there more details about JobController interface and backend implementation design?

Reviewer: There isn't much documentation on the backend implementation design yet. The [README.md](https://github.com/kubeflow/common/blob/master/README.md) and the comments in [the interface](https://github.com/kubeflow/common/blob/master/pkg/apis/common/v1/interface.go) contains some information.

-   [Background](#background)
-   [Motivation](#motivation)
-   [Goals](#goals)
-   [Non-Goals](#non-goals)
-   [Design](#design)
    -   [Container images and deploying FATE cluster on
        Kubernetes](#container-images-and-deploying-fate-cluster-on-kubernetes)
    -   [Custom Resource Definition](#custom-resource-definition)
        -   [Kubefate](#kubefate)
        -   [FATECluster](#fatecluster)
        -   [FateJob](#fatejob)
    -   [Controller design](#controller-design)
-   [Reference](#reference)

_Status_
* 2020-5-28 - Draft v2
* 2020-5-14 – Draft v1

## Background
Federated machine learning (FML) is a machine learning setting where many clients (e.g. mobile devices or organizations) collaboratively train a model under the coordination of a central server while keeping the training data decentralized. Only the encrypted mediate parameters are exchanged between clients with MPC or homomorphic encryption.
![Federated Machine Learning](diagrams/fate-operator-fl.png)

> FML has received significant interest recently, because of its effectiveness to solve data silos and data privacy preserving problems. Companies participated in federated machine learning include 4Paradigm, ANT Financial, Data Republic, Google, Huawei, Intel, JD.com, Microsoft, Nvidia, OpenMind, Pingan Technology, Sharemind, Tencent, VMware, Webank etc. 

Reviewer 3: Companies doing research and building their technology solutions around federated machine learning include 4Paradigm, ANT Financial, Data Republic, Google, Huawei, Intel, JD.com, Microsoft, Nvidia, OpenMind, Pingan Technology, Sharemind, Tencent, VMware, Webank etc.

Reviewer: ANT Financial -> Ant Financial

Reviewer 2: @terrytangyuan BTW, is there any plan for Ant Financial to open source its internal FML framework?

Reviewer: Not that I am aware of.

> Depending on the differences in features and sample data space, federated machine learning can be classified into horizontally federated machine learning, vertically federated machine learning and federated transfer learning. Horizontally federated machine learning also is called sample-based federated machine learning, means data sets share the same feature space but different in samples. With horizontally federated machine learning, we can gather the relatively small or partial data set into a big one to in-crease the performance of trained models. Vertical federated machine learning is applicable to the cases that two data set with different feature space but share same sample ID. With vertical federated machine learning we can train a model with attributes from different organizations for a full profile. Vertical federated machine learning is required to redesign most of machine learning algorithms. Federated transfer learning applies to scenarios that two data set with different features space but also different samples. 

Reviewer 3: ```suggestion
Depending on the differences in features and sample data space, federated machine learning can be classified into _horizontal federated machine learning_, _vertical federated machine learning_ and _federated transfer learning_. Horizontal federated machine learning is also called sample-based federated machine learning, which means data sets share the same feature space but have different samples. With horizontal federated machine learning, we can gather the relatively small or partial datasets into a big one to increase the performance of trained models. Vertical federated machine learning is applicable to the cases where there are two datasets with different feature space but share same sample ID. With vertical federated machine learning we can train a model with attributes from different organizations for a full profile. Vertical federated machine learning is required to redesign most of machine learning algorithms. Federated transfer learning applies to scenarios where there are two datasets with different features space but also different samples. 
```

Author: Thanks for the update suggestions, have merge in latest commit.

   * Kubefate: the resource management component of FATE cluster.
2.	Example of full lifecycle to create KubeFATE component, deploy FATE cluster and submit an FML job to created FATE and get the result. Note that, KubeFATE and FATE cluster only needs to be deployed once, and can handle multiple jobs.

## Non-Goals
For the scope of this proposal, we won’t be addressing the method of serving the model.

## Design

### Container images and deploying FATE cluster on Kubernetes
We have built a set of Docker images for FATE cluster, and put into: https://hub.docker.com/orgs/federatedai/repositories . All the images have been already used and verified by community users. 

There is a provisioning and management component of FATE cluster, called [KubeFATE](https://github.com/FederatedAI/KubeFATE/tree/master/k8s-deploy). KubeFATE manages FATE clusters of one party in a federation.

All images work well and are proven in users’ environments.

### Custom Resource Definition
#### Kubefate
```
apiVersion: app.kubefate.net/v1beta1
kind: Kubefate
metadata:
  name: kubefate-sample
  namespace: kube-fate
spec:
  # kubefate image tag
  imageVersion: v1.0.2
  # ingress host

>   ingressDomain: kubefate.net

Reviewer 4: Is Kubefate creating the ingress and setting up networking? The architecture in Kubeflow is to use ISTIO to create a service mesh. Different platforms can then expose the external endpoint as makes sense on their platform.

Author: There are three ingress will exposed: 1. The KubeFATE service will exposed for serving management requests; 2. The FATE-Board component of FATE; 3. The Proxy to connect with other organizations. The first two have been changed to istio gateway. The proxy now is expose with nodeport and connect with other organization. We are working on refactor it to expose gateway now. It will take some times.

  # kubefate config
  config:

>     FATECLOUD_MONGO_USERNAME: "root"

Reviewer 4: Are alternative databases to MONGO possible? I think MONGO changed their license and is no longer a widely accepted OSS license like Apache2.  This might limit how closely we can integrate the two. For example, it might prevent us from including Fate in our standard configurations or hosting those images in our repos.

Also is it possible to use secrets to store the password and not put it in the config. 

/cc @animeshsingh

Author: Yes. It has been changed to secrets. I will update the CRD sample in proposal right now.

Reviewer 3: @LaynePeng Apache CouchDB can be a more elegant choice, given its already in Apache foundation
https://couchdb.apache.org/
